README.md
# Fluentd Server

[![Build Status](https://secure.travis-ci.org/sonots/fluentd-server.png?branch=master)](http://travis-ci.org/sonots/fluentd-server)
[![Coverage Status](https://coveralls.io/repos/sonots/fluentd-server/badge.png?branch=master)](https://coveralls.io/r/sonots/fluentd-server?branch=master)

A Fluentd config distribution server

Demo: [http://fluentd-server.herokuapp.com](http://fluentd-server.herokuapp.com)

## What You Can Do

With Fluentd Server, you can manage fluentd configuration files centrally with `erb`. 

For example, you may create a config post whose name is `worker` as:

```
<source>
  type forward
  port <%= port %>
</source>

<match **>
  type stdout
</match>
```

Then you can download the config via an API whose uri is like `/api/worker?port=24224` where its query parameters are replaced with variables in the erb. 
The downloaded contents should become as follows:

```
<source>
  type forward
  port 24224
</source>

<match **>
  type stdout
</match>
```

## How to Use

The `include` directive of fluentd config supports `http`, so write just one line on your fluentd.conf as:

```
# /etc/fluentd.conf
include http://fqdn.to.fluentd-server/api/:name?port=24224
```

so that it will download the real configuration from the Fluentd Server where :name is the name of your post. 

## Installation

Prerequisites

* SQLite
* Ruby 2.0 or later

### From Gem package

Easy steps on installation with gem and SQLite.

```bash
$ gem install fluentd-server
$ gem install sqlite3
$ fluentd-server new
$ cd fluentd-server
$ fluentd-server init # creates database scheme on SQLite
$ fluentd-server start
```

Then see `http://localhost:5126/`.

### From Git repository

Install from git repository. 

```bash
$ git clone https://github.com/sonots/fluentd-server.git
$ cd fluentd-server
$ bundle
$ bundle exec fluentd-server init # creates database scheme on SQLite
$ bundle exec fluentd-server start
```

Then see `http://localhost:5126/`. 

## Configuration

To configure fluentd-server, edit the `.env` file in the project root directory.

The default configuration is as follows:

```
PORT=5126
HOST=0.0.0.0
# DATABASE_URL=sqlite3:data/fluentd_server.db
# JOB_DIR=jobs
# LOG_PATH=STDOUT
# LOG_LEVEL=debug
# LOG_SHIFT_AGE=0
# LOG_SHIFT_SIZE=1048576
```

## HTTP API

See [API.md](API.md).

### Use Fluentd Server from Command Line

For the case you want to edit Fluentd configuration files from your favorite editors rather than from the Web UI, `FILE STORAGE` feature is available.
With this feature, you should also be able to manage your configuration files with git (or any VCS).

To use this feature, enable `FILE_STORAGE` in `.env` file as:

```
FILE_STORAGE=true
DATA_DIR=data
SYNC_INTERVAL=60
```

where the `DATA_DIR` is the location to place your configuration files locally, and the `SYNC_INTERVAL` is the interval where a synchronization worker works.

Place your `erb` files in the `DATA_DIR` directory, and please execute `sync` command to synchronize the file existence information with DB
when you newly add or remove the configuration files.

```
$ fluentd-server sync
```

Or, you may just wait `SYNC_INTERVAL` senconds until a synchronization worker automatically synchronizes the information.
Please note that modifying the file content does not require to synchronize because the content is read from the local file directly.

NOTE: Enabling this feature disables to edit the Fluentd configuration from the Web UI.

### CLI (Command Line Interface)

Here is a full list of fluentd-server commands.

```bash
$ fluentd-server help
Commands:
  fluentd-server help [COMMAND]        # Describe available commands or one specific command
  fluentd-server init                  # Creates database schema
  fluentd-server job-clean             # Clean fluentd_server delayed_job queue
  fluentd-server job-worker            # Sartup fluentd_server job worker
  fluentd-server migrate               # Migrate database schema
  fluentd-server new                   # Creates fluentd-server resource directory
  fluentd-server start                 # Sartup fluentd_server
  fluentd-server sync                  # Synchronize local file storage with db immediately
  fluentd-server sync-worker           # Sartup fluentd_server sync worker
  fluentd-server td-agent-condrestart  # Run `/etc/init.d/td-agent condrestart` via serf event
  fluentd-server td-agent-configtest   # Run `/etc/init.d/td-agent configtest` via serf query
  fluentd-server td-agent-reload       # Run `/etc/init.d/td-agent reload` via serf event
  fluentd-server td-agent-restart      # Run `/etc/init.d/td-agent restart` via serf event
  fluentd-server td-agent-start        # Run `/etc/init.d/td-agent start` via serf event
  fluentd-server td-agent-status       # Run `/etc/init.d/td-agent status` via serf query
  fluentd-server td-agent-stop         # Run `/etc/init.d/td-agent stop` via serf event
```

## ToDo

* Automatic deployment (restart) support like the one of chef-server

  * Need a notification function for when configtest or restart failed
  * Pipe the resulted json to a command (I may prepare email.rb as an example).

## ChangeLog

See [CHANGELOG.md](CHANGELOG.md) for details.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](../../pull/new/master)

## Copyright

Copyright (c) 2014 Naotoshi Seo. See [LICENSE](LICENSE) for details.


CHANGELOG.md
# 0.3.2 (2014/06/08)

Changes:

* Rename `LOCAL_STORAGE` to `FILE_STORAGE`

# 0.3.1

Enhancements:

* Add `fluentd-server sync` command
* Add `fluentd-server td-agent-start` command
* Add `fluentd-server td-agent-stop` command
* Add `fluentd-server td-agent-reload` command
* Add `fluentd-server td-agent-restart` command
* Add `fluentd-server td-agent-condrestart` command
* Add `fluentd-server td-agent-status` command
* Add `fluentd-server td-agent-configtest` command

Changes:

* Rename `fluentd-server sync` command to `fluentd-server sync-worker`
* Rename `fluentd-server job` command to `fluentd-server job-worker`
* Rename `fluentd-server job_clear` command to `fluentd-server job-clean`

# 0.3.0

Enhancements:

* Add sync_worker

Changes:

* Use env LOCAL_STORAGE rather than DATA_DIR to enable local file storage feature

# 0.2.0

Enhancements:

* Support restarting td-agent via serf
  * Created serf-td-agent gem
  * Introduce Resque
  * Ajax reload of command result
* Save/load the content of configuration into/from a file (experimental)
  * Created acts_as_file gem

# 0.1.0

First version



lib/fluentd_server/cli.rb
require "fileutils"
require "dotenv"
require "thor"
require 'fluentd_server'

class FluentdServer::CLI < Thor
  BASE_DIR = File.join(Dir.pwd, "fluentd-server")
  DATA_DIR = File.join(BASE_DIR, "data")
  LOG_DIR = File.join(BASE_DIR, "log")
  JOB_DIR = File.join(BASE_DIR, "jobs")
  LOG_FILE = File.join(LOG_DIR, "application.log")
  ENV_FILE = File.join(BASE_DIR, ".env")
  PROCFILE = File.join(BASE_DIR, "Procfile")
  CONFIGRU_FILE = File.join(BASE_DIR, "config.ru")
  CONFIG_DIR= File.join(BASE_DIR, "config")

  DEFAULT_DOTENV =<<-EOS
PORT=5126
HOST=0.0.0.0
DATABASE_URL=sqlite3:#{DATA_DIR}/fluentd_server.db
JOB_DIR=#{JOB_DIR}
LOG_PATH=#{LOG_FILE}
LOG_LEVEL=warn
LOG_SHIFT_AGE=0
LOG_SHIFT_SIZE=1048576
FILE_STORAGE=false
DATA_DIR=#{DATA_DIR}
SYNC_INTERVAL=60
EOS

  DEFAULT_PROCFILE =<<-EOS
web: unicorn -E production -p $PORT -o $HOST -c config/unicorn.conf
job: fluentd-server job-worker
sync: fluentd-server sync-worker
serf: $(gem path serf-td-agent)/bin/serf agent
EOS

  default_command :start

  def initialize(args = [], opts = [], config = {})
    super(args, opts, config)
  end

  desc "new", "Creates fluentd-server resource directory"
  def new
    FileUtils.mkdir_p(LOG_DIR)
    FileUtils.mkdir_p(JOB_DIR)
    File.write ENV_FILE, DEFAULT_DOTENV
    File.write PROCFILE, DEFAULT_PROCFILE
    FileUtils.cp(File.expand_path("../../../config.ru", __FILE__), CONFIGRU_FILE)
    FileUtils.cp_r(File.expand_path("../../../config", __FILE__), CONFIG_DIR)
    puts 'fluentd-server new finished.'
  end

  desc "init", "Creates database schema"
  def init
    Dotenv.load
    require 'fluentd_server/environment'
    require 'rake'
    require 'sinatra/activerecord/rake'
    ActiveRecord::Tasks::DatabaseTasks.db_dir = File.expand_path("../../../db", __FILE__)
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths = [File.expand_path("../../../db/migrate", __FILE__)]
    # ToDo: Fix that db:migrate raises an error in the case of sqlite3 like
    # SQLite3::SQLException: database schema has changed: INSERT INTO "schema_migrations" ("version") VALUES (?)
    # Rake::Task['db:migrate'].invoke
    # Use db:schema:load after generating db/schema.rb by executing db:migrate several times for now
    Rake::Task['db:schema:load'].invoke
    puts 'fluentd-server init finished.'
  end

  desc "migrate", "Migrate database schema"
  def migrate
    Dotenv.load
    require 'fluentd_server/environment'
    require 'rake'
    require 'sinatra/activerecord/rake'
    ActiveRecord::Tasks::DatabaseTasks.db_dir = File.expand_path("../../../db", __FILE__)
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths = [File.expand_path("../../../db/migrate", __FILE__)]
    Rake::Task['db:migrate'].invoke
  end

  desc "start", "Sartup fluentd_server"
  def start
    self.migrate # do migration before start not to forget on updation
    require "foreman/cli"
    Foreman::CLI.new.invoke(:start, [], {})
  end

  # reference: https://gist.github.com/robhurring/732327
  desc "job-worker", "Sartup fluentd_server job worker"
  def job_worker
    Dotenv.load
    require 'delayed_job'
    require 'fluentd_server/model'
    worker_options = {
      :min_priority => ENV['MIN_PRIORITY'],
      :max_priority => ENV['MAX_PRIORITY'],
      :queues => (ENV['QUEUES'] || ENV['QUEUE'] || '').split(','),
      :quiet => false
    }
    Delayed::Worker.new(worker_options).start
  end

  desc "job-clean", "Clean fluentd_server delayed_job queue"
  def job_clean
    Dotenv.load
    require 'delayed_job'
    require 'fluentd_server/model'
    Delayed::Job.delete_all
  end

  desc "sync-worker", "Sartup fluentd_server sync worker"
  def sync_worker
    Dotenv.load
    require 'fluentd_server/sync_worker'
    FluentdServer::SyncWorker.start
  end

  desc "sync", "Synchronize local file storage with db immediately"
  def sync
    Dotenv.load
    require 'fluentd_server/sync_runner'
    FluentdServer::SyncRunner.run
  end

  desc "td-agent-start", "Run `/etc/init.d/td-agent start` via serf event"
  def td_agent_start
    Dotenv.load
    require 'fluentd_server/model'
    system("#{::Task.serf_path} event td-agent-start")
  end

  desc "td-agent-stop", "Run `/etc/init.d/td-agent stop` via serf event"
  def td_agent_stop
    Dotenv.load
    require 'fluentd_server/model'
    system("#{::Task.serf_path} event td-agent-stop")
  end

  desc "td-agent-reload", "Run `/etc/init.d/td-agent reload` via serf event"
  def td_agent_reload
    Dotenv.load
    require 'fluentd_server/model'
    system("#{::Task.serf_path} event td-agent-reload")
  end

  desc "td-agent-restart", "Run `/etc/init.d/td-agent restart` via serf event"
  def td_agent_restart
    Dotenv.load
    require 'fluentd_server/model'
    # ::Task.create_and_delete(name: 'Restart').restart # using delayed_job
    system("#{::Task.serf_path} event td-agent-restart")
  end

  desc "td-agent-condrestart", "Run `/etc/init.d/td-agent condrestart` via serf event"
  def td_agent_condrestart
    Dotenv.load
    require 'fluentd_server/model'
    system("#{::Task.serf_path} event td-agent-condrestart")
  end

  desc "td-agent-status", "Run `/etc/init.d/td-agent status` via serf query"
  def td_agent_status
    Dotenv.load
    require 'fluentd_server/model'
    # ::Task.create_and_delete(name: 'Status').status # using delayed_job
    system("#{::Task.serf_path} query td-agent-status")
  end

  desc "td-agent-configtest", "Run `/etc/init.d/td-agent configtest` via serf query"
  def td_agent_configtest
    Dotenv.load
    require 'fluentd_server/model'
    # ::Task.create_and_delete(name: 'Configtest').configtest # using delayed_job
    system("#{::Task.serf_path} query td-agent-configtest")
  end

  no_tasks do
    def abort(msg)
      $stderr.puts msg
      exit 1
    end
  end
end
